src/gridcharger/huawei/HardwareInterface.cpp

```diff
@@ -51,111 +51,458 @@ void HardwareInterface::stopLoop()
+        if (config.Huawei.VerboseLogging) {
+            MessageOutput.printf("[Huawei::HwIfc] max current multiplier is %.2f\r\n",
+                    _maxCurrentMultiplier);
+        }
+    }
+    else if (counter == 6) {
+        _upData->add<DataPointLabel::Row>(static_cast<uint8_t>((msg.valueId >> 8) & 0xFF));
+        _upData->add<DataPointLabel::Slot>(static_cast<uint8_t>(msg.valueId & 0xFF));
+        _upData->add<DataPointLabel::Reachable>(true, true/*ignore age*/);
+
+        _lastDeviceConfigMillis = millis();
+    }
+
+    return true;
+}
+
+void HardwareInterface::requestDeviceConfig()
+{
+    _sendQueue.push(command_t {
+        .tries = 1,
+        .deviceAddress = 1,
+        .registerAddress = 0x50FE,
+        .command = 0,
+        .flags = 0,
+        .value = 0
+    });
+}
+
+bool HardwareInterface::readRectifierState(can_message_t const& msg)
+{
+    auto const& config = Configuration.get();
+
+    // we will receive a bunch of messages with CAN ID 0x1081407F,
+    // and one (the last one) with ID 0x1081407E.
+    if ((msg.canId | 0x1) != 0x1081407F) { return false; }
+
+    uint32_t valueId = msg.valueId;
+
+    // sometimes the last bit of the value ID of a message with CAN ID
+    // 0x1081407E is set. TODO(schlimmchen): why?
+    if (msg.canId == 0x1081407E && (valueId & 0x01) > 0) {
+        if (config.Huawei.VerboseLogging) {
+            MessageOutput.printf("[Huawei::HwIfc] last bit in value ID %08x is "
+                    "set, resetting\r\n", valueId);
+        }
+        valueId &= ~(0x01);
+    }
+
+    // for unknown reasons, the input voltage value ID has the last two bits
+    // set on a R4830G1. this unit supports DC input as well, but these bits
+    // do not change when powered the unit using DC. TODO(schlimmchen): why?
+    if (msg.canId == 0x1081407F && (valueId & 0x03) > 0) {
+        if (config.Huawei.VerboseLogging) {
+            MessageOutput.printf("[Huawei::HwIfc] last two bits in value ID "
+                    "%08x are set, resetting\r\n", valueId);
+        }
+        valueId &= ~(0x03);
+    }
+
+    // during start-up and when shortening or opening the slot detect pins,
+    // the value ID starts with 0x31 rather than 0x01. TODO(schlimmchen): why?
+    if ((valueId >> 24) == 0x31) {
+        if (config.Huawei.VerboseLogging) {
+            MessageOutput.print("[Huawei::HwIfc] processing value for value ID "
+                    "starting with 0x31\r\n");
+        }
+        valueId &= 0x0FFFFFFF;
+    }
+
+    if ((valueId & 0xFF00FFFF) != 0x01000000) { return false; }
+
+    auto label = static_cast<DataPointLabel>((valueId & 0x00FF0000) >> 16);
+
+    float divisor = (label == DataPointLabel::OutputCurrentMax) ? _maxCurrentMultiplier : 1024;
+
+    if (divisor == 0) {
+        MessageOutput.print("[Huawei::HwIfc] cannot process output current max "
+                "value while respective multiplier unknown\r\n");
+        return false;
+    }
```

Comment:
~~Will the divisor actually ever be 0?~~

~~We either use _maxCurrentMultiplier or 1024~~

Now i get it.

The below change would make it a bit easier to understand
```suggestion
    if (_maxCurrentMultiplier == 0 && label == DataPointLabel::OutputCurrentMax) {
        MessageOutput.print("[Huawei::HwIfc] cannot process output current max "
                "value while respective multiplier unknown\r\n");
        return false;
    }

    float divisor = (label == DataPointLabel::OutputCurrentMax) ? _maxCurrentMultiplier : 1024;
```

Reply:
Hm. If you want this more explicit, okay. But then let's be really explicit and don't repeat the test for the value of label.

Reply:
See e81d22280795b8f332697846458a5df1a04949f0